resqpy/olio/weights_and_measures.py

```diff
@@ -58,27 +75,59 @@ def rq_uom(units):
    if ul in ['1E6 m3/d', '1E6 m3/day']: return '1E6 m3/d'
    if units in ['mD.m', 'mD.ft']: return units
    if ul == 'count': return 'Euc'
-   uom_list = properties_data()['uoms']
-   if units in uom_list: return units
    if ul in uom_list: return ul  # dangerous! for example, 'D' means D'Arcy and 'd' means day
    return 'Euc'
 
 
-def rq_uom_list(units_list):
-   """Returns a list of RESQML uom equivalents for units in list."""
+def convert(x, unit_from, unit_to):
+   """Convert value between two compatible units
+   
+   Args:
+      x (numeric or np.array): value(s) to convert
+      unit_from (str): resqml uom
+      unit_to (str): resqml uom
 
-   rq_list = []
-   for u in units_list: rq_list.append(rq_uom(u))
-   return rq_list
+   Returns:
+      Converted value(s)
 
+   """
+   # TODO: robust handling of errors. At present, bad units are treated as "EUC", and will fail silently.
 
-def p_length_unit(units):
-   """Returns length units string as expected by pagoda weights and measures module."""
+   # conversion data assume the formula "y=(A + Bx)/(C + Dx)" where "y" represents a value in the base unit.
+   # Backwards formula: x=(A-Cy)/(Dy-B)
+   # All current units have D==0
 
-   # NB: other length units are supported by resqml
-   if units.lower() in ['m', 'metre', 'metres']: return 'metres'
-   if units.lower() in ['ft', 'foot', 'feet', 'ft[us]']: return 'feet'
-   assert(False)  # unrecognised length units
+   A1, B1, C1, D1 = get_conversion_factors(rq_uom(unit_from))
+   A2, B2, C2, D2 = get_conversion_factors(rq_uom(unit_to))
+   y = (A1 + (B1*x)) / (C1 + (D1*x))
```

Comment:
Something missing: this function should check that the units actually share the same base units, else the conversion will be nonsense.

The function 'get_conversion_factors' could perhaps also return the base unit.

[This is Connor on mobile]

Reply:
Yes, agreed it would be good to do a check here – and bundling the base units into the results of get_conversion_factors() seems appropriate. Note that this convert function needs to be fast.

I would also add a line along the lines of _if unit_from == unit_to: return x_